src/main/scala/uk/gov/nationalarchives/dp/client/Client.scala

```diff
@@ -0,0 +1,139 @@
+package uk.gov.nationalarchives.dp.client
+
+import cats.MonadError
+import cats.implicits._
+import sttp.capabilities.Streams
+import sttp.client3._
+import sttp.client3.upicklejson._
+import uk.gov.nationalarchives.dp.client.Client.{AuthDetails, BitStreamInfo, Entity}
+import upickle.default._
+
+import java.time.format.DateTimeFormatter
+import java.time.ZonedDateTime
+import java.util.UUID
+import scala.xml.{Elem, XML}
+
+trait Client[F[_], S] {
+  def metadataForEntityUrl(url: String, auth: AuthDetails): F[Seq[Elem]]
+
+  def getBitstreamInfo(contentId: UUID, authDetails: AuthDetails): F[Seq[BitStreamInfo]]
+
+  def streamBitstreamContent[T](
+      stream: Streams[S]
+  )(url: String, authDetails: AuthDetails, streamFn: stream.BinaryStream => F[T]): F[T]
+
+  def entitiesUpdatedSince(dateTime: ZonedDateTime, authDetails: AuthDetails): F[Seq[Entity]]
+}
+
+object Client {
+  private val asXml: ResponseAs[Either[String, Elem], Any] = asString.mapRight(XML.loadString)
+  implicit val responsePayloadRW: ReadWriter[Token] = macroRW[Token]
+
+  case class AuthDetails(userName: String, password: String)
+
+  case class Token(token: String)
+
+  case class BitStreamInfo(name: String, url: String)
+
+  case class Entity(ref: String, title: String, entityType: String, url: String, deleted: Boolean)
+
+  def createClient[F[_], S](apiBaseUrl: String, backend: SttpBackend[F, S])(implicit
+      me: MonadError[F, Throwable]
+  ): Client[F, S] = new Client[F, S] {
+    val dataProcessor: DataProcessor[F] = DataProcessor[F]()
+
+    implicit class EitherUtils[T](e: Either[String, T]) {
+      def bodyLift: F[T] = me.fromTry(e.left.map(err => new RuntimeException(err)).toTry)
+    }
+
+    private def getAuthenticationToken(authDetails: AuthDetails): F[String] = {
+      val response = basicRequest
+        .body(Map("username" -> authDetails.userName, "password" -> authDetails.password))
+        .post(uri"$apiBaseUrl/api/accesstoken/login")
+        .response(asJson[Token])
+        .send(backend)
+      me.flatMap(response)(res => me.fromTry(res.body.map(b => b.token).toTry))
```

Comment:
1. There are quite a few brackets, maybe one could be a set of braces for clarity
```
me.flatMap(response){
  res => me.fromTry(res.body.map(b => b.token).toTry)
}
```

2.  I could be wrong, but I think that the "b" in `b => b.token` in actually a token object ,maybe, it can be replaced with "t"? or just `_.token`?

Reply:
You're right about the brackets, that looks better. And I've changed tbe b to a t as well.

Reply:
You're right about the brackets, that looks better. And I've changed tbe b to a t as well.